tests/Benchmarks/System.Text.JsonLab/JsonReaderPerf.cs

```diff
@@ -2,70 +2,277 @@
+                case TestCaseType.HelloWorld:
+                    _jsonString = JsonStrings.HelloWorld;
+                    break;
+                case TestCaseType.Basic:
+                    _jsonString = JsonStrings.BasicJson;
+                    break;
+                case TestCaseType.BasicLargeNum:
+                    _jsonString = JsonStrings.BasicJsonWithLargeNum;
+                    break;
+                case TestCaseType.SpecialNumForm:
+                    _jsonString = JsonStrings.JsonWithSpecialNumFormat;
+                    break;
+                case TestCaseType.ProjectLockJson:
+                    _jsonString = JsonStrings.ProjectLockJson;
+                    break;
+                case TestCaseType.FullSchema1:
+                    _jsonString = JsonStrings.FullJsonSchema1;
+                    break;
+                case TestCaseType.FullSchema2:
+                    _jsonString = JsonStrings.FullJsonSchema2;
+                    break;
+                case TestCaseType.DeepTree:
+                    _jsonString = JsonStrings.DeepTree;
+                    break;
+                case TestCaseType.BroadTree:
+                    _jsonString = JsonStrings.BroadTree;
+                    break;
+                case TestCaseType.LotsOfNumbers:
+                    _jsonString = JsonStrings.LotsOfNumbers;
+                    break;
+                case TestCaseType.LotsOfStrings:
+                    _jsonString = JsonStrings.LotsOfStrings;
+                    break;
+                case TestCaseType.Json400B:
+                    _jsonString = JsonStrings.Json400B;
+                    break;
+                case TestCaseType.Json4KB:
+                    _jsonString = JsonStrings.Json4KB;
+                    break;
+                case TestCaseType.Json40KB:
+                    _jsonString = JsonStrings.Json40KB;
+                    break;
+                case TestCaseType.Json400KB:
+                    _jsonString = JsonStrings.Json400KB;
+                    break;
+            }
 
-            _stream = new MemoryStream(_data);
-            var enc = Target == EncoderTarget.InvariantUtf8 ? Encoding.UTF8 : Encoding.Unicode;
-            _reader = new StreamReader(_stream, enc, false, 1024, true);
+            if (IsUTF8Encoded)
+            {
+                _dataUtf8 = Encoding.UTF8.GetBytes(_jsonString);
+                _stream = new MemoryStream(_dataUtf8);
+                _reader = new StreamReader(_stream, Encoding.UTF8, false, 1024, true);
+            }
+            else
+            {
+                _dataUtf16 = Encoding.Unicode.GetBytes(_jsonString);
+            }
         }
 
         [Benchmark(Baseline = true)]
-        public void ReaderNewtonsoftStringReader()
+        public void ReaderNewtonsoftReaderEmptyLoop()
         {
-            var json = new Newtonsoft.Json.JsonTextReader(new StringReader(JsonString));
-            while (json.Read()) ;
+            if (IsUTF8Encoded)
```

Comment:
I think that it would be nice to simplify the code by reducing the repeated code, an example:

```cs
_stream.Seek(0, SeekOrigin.Begin);

var reader = IsUTF8Encoded ? _reader : new StringReader(_jsonString);
var json = new Newtonsoft.Json.JsonTextReader(reader);
while (json.Read()) ;
```

Reply:
I don't want to incur the cost of resetting the stream in all cases, but I will reduce the duplicated code.

```C#
TextReader reader;
if (IsUTF8Encoded)
{
    _stream.Seek(0, SeekOrigin.Begin);
    reader = _reader;
}
else
{
    reader = new StringReader(_jsonString);
}
var json = new Newtonsoft.Json.JsonTextReader(reader);
while (json.Read()) ;
```